fix: handle empty Authorization headers in API auth strategy

Vansh5632 opened the pull request:
## What does this PR do?

- Treat empty Authorization headers ('') same as missing headers
- Fixes 401 errors when external services send empty auth headers
- Resolves ElevenLabs book_meeting API integration issue
- Maintains backward compatibility with existing auth methods

- Fixes #23683  (GitHub issue number)
- Fixes CAL- 6383(Linear issue number - should be visible at the bottom of the GitHub issue description)

#### Video Demo (if applicable):


https://github.com/user-attachments/assets/57938364-9e84-403f-a13e-f80ec1f01575

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?
tested using this curl request 
```code
curl -X POST "http://localhost:3002/v2/bookings" \
  -H "cal-api-version: 2024-08-13" \
  -H "Authorization: " \
  -H "Content-Type: application/json" \
  -d '{
    "eventTypeId": 1,
    "start": "2024-09-15T10:00:00Z",
    "responses": {
      "name": "Test User from ElevenLabs",
      "email": "test@elevenlabs.com"
    }
  }' \
  -v
```

<!-- devin-review-badge-begin -->

---

<a href="https://app.devin.ai/review/calcom/cal.com/pull/23824">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    <img src="https://static.devin.ai/assets/gh-open-in-devin-review-light.svg?v=1" alt="Open with Devin">
  </picture>
</a>
<!-- devin-review-badge-end -->

Vansh5632 committed:
fix: handle empty Authorization headers in API auth strategy

- Treat empty Authorization headers ('') same as missing headers
- Fixes 401 errors when external services send empty auth headers
- Resolves ElevenLabs book_meeting API integration issue
- Maintains backward compatibility with existing auth methods

Devanshusharma2005 committed:
Merge branch 'main' into fix/empty-authorization-header-handling

Vansh5632 committed:
fix: make Authorization header parsing case-insensitive and robust

- Replace simple string replace with regex matching /^\s*Bearer\s+(.+)\s*$/i
- Handle case-insensitive Bearer scheme matching
- Properly handle extra spaces around Bearer and token
- Treat zero-length tokens as absent (undefined)
- Requests like 'Authorization: Bearer   ' now fall back same as missing header

Vansh5632 committed:
Merge branch 'main' into fix/empty-authorization-header-handling

pedroccastro reviewed (dismissed):
Thanks @Vansh5632 for your contribution, really appreciate this! 🙌

The changes look good to me. The case-insensitive Bearer scheme matching is correct, and the handling of empty/whitespace Authorization headers is solid and well thought out.

One request: Could you add unit tests covering the edge cases this fix addresses? Specifically:

```
Authorization: "" (empty string)
Authorization: " " (whitespace only)
Authorization: "Bearer " (Bearer without token)
Authorization: "bearer token" (lowercase scheme)
```

This will help prevent regressions and document the expected behavior

**Non-blocking follow-up suggestion:** There are a few other places in the codebase still using the .replace("Bearer ", "") pattern (for example, permissions.guard.ts and throttler-guard.ts). It might be worth extracting this logic into a small shared utility to keep the behavior consistent and avoid subtle discrepancies over time

Vansh5632 committed:
feat: add unit tests and extract Bearer token utility

- Add comprehensive unit tests for Bearer token edge cases
  - Test empty string Authorization headers
  - Test whitespace-only Authorization headers
  - Test 'Bearer ' with no token
  - Test lowercase 'bearer' scheme
  - Additional edge cases for robustness

- Create shared extractBearerToken() utility
  - Case-insensitive Bearer scheme matching
  - Handles empty/whitespace headers gracefully
  - Properly validates token presence
  - Exported from @/lib/api-key for consistency

- Refactor 13 files to use shared utility
  - Replace all .replace('Bearer ', '') with extractBearerToken()
  - Ensures consistent behavior across codebase
  - Prevents subtle discrepancies in auth handling

Addresses maintainer feedback on PR #23824

Vansh5632 committed:
Merge branch 'main' into fix/empty-authorization-header-handling

Vansh5632 committed:
fix: validate extractBearerToken result to prevent undefined token in OAuth state

- Add undefined check after extractBearerToken() in gcal.controller.ts
- Add undefined check after extractBearerToken() in stripe.controller.ts
- Add undefined check after extractBearerToken() in organizations-stripe.controller.ts
- Add undefined check after extractBearerToken() in organizations-conferencing.controller.ts
- Add e2e tests for empty Authorization header scenarios in gcal.controller.e2e-spec.ts

Throws UnauthorizedException when no valid Bearer token is found, preventing
"undefined" from being stringified into OAuth state parameters.

devin-ai-integration[bot] committed:
Merge branch 'main' from upstream into fix/empty-authorization-header-handling

Co-Authored-By: unknown <>

dhairyashiil committed:
fix

dhairyashiil reviewed (dismissed):
LGTM

dhairyashiil committed:
Merge branch 'main' into fix/empty-authorization-header-handling

devin-ai-integration[bot] committed:
Merge branch 'main' from upstream into fix/empty-authorization-header-handling

Co-Authored-By: unknown <>

sahitya-chandra reviewed (dismissed):
LGTM

sahitya-chandra committed:
Merge branch 'main' into fix/empty-authorization-header-handling

sahitya-chandra committed:
Validate access token in gcal.service.ts

Ensure accessToken is validated before proceeding.

sahitya-chandra committed:
Require valid Bearer token in outlook service

Throw UnauthorizedException if access token is missing.

ThyMinimalDev committed:
Merge branch 'main' into fix/empty-authorization-header-handling

Vansh5632 committed:
Merge branch 'main' into fix/empty-authorization-header-handling

Vansh5632 commented:
@keithwillcode Sorry about that. I saw the mail notification and assumed it was required. I’ll keep this in mind going forward.

keithwillcode commented:
> @keithwillcode Sorry about that. I saw the mail notification and assumed it was required. I’ll keep this in mind going forward.

No worries at all. Appreciate the PR! Trying to get it merged ASAP. Failing test...seeing if flaky

PeerRich committed:
Merge branch 'main' into fix/empty-authorization-header-handling

anikdhabal committed:
Merge branch 'main' into fix/empty-authorization-header-handling

romitg2 committed:
Merge branch 'main' into fix/empty-authorization-header-handling

sahitya-chandra reviewed (commented):
Thank you for your contribution. This PR is being closed because it modifies code that no longer exists in Cal.diy
